File: javascript/websearch-crawl.js
#!/usr/bin/env node
// websearch-crawl.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway1:scripts/websearch-crawl.sh
// auch in: OpenClaw@gateway2:scripts/websearch-crawl.sh
// Erzeugt: 2026-08-23 durch ABSTRACTIONS_MANAGER.py

import fs from 'fs';
import path from 'path';
import https from 'https';
import { spawn } from 'child_process';

// Web Search Script: Website Crawling mit Firecrawl
// Verwendung: node websearch-crawl.js <URL> [OUTPUT_DIR]

const args = process.argv.slice(2);
const WEBSITE_URL = args[0];
const OUTPUT_DIR = args[1] || './crawled';

// API-Key aus Umgebungsvariable oder aus Datei lesen
let FIRECRAWL_API_KEY = process.env.FIRECRAWL_API_KEY;
if (!FIRECRAWL_API_KEY) {
  try {
    const openclawEnvPath = path.join(process.env.HOME, '.openclaw', 'openclaw.env');
    const envContent = fs.readFileSync(openclawEnvPath, 'utf-8');
    const match = envContent.match(/OPENROUTER\s*=\s*"([^"]+)"/);
    if (match) {
      FIRECRAWL_API_KEY = match[1];
    }
  } catch (err) {
    // Datei nicht gefunden oder anderer Fehler - API_KEY bleibt undefined
  }
}

if (!WEBSITE_URL) {
  console.log('Verwendung: node websearch-crawl.js <URL> [OUTPUT_DIR]');
  process.exit(1);
}

// OUTPUT_DIR erstellen
fs.mkdirSync(OUTPUT_DIR, { recursive: true });
console.log(`Crawling ${WEBSITE_URL}...`);

// Hilfsfunktion für HTTP-Anfragen
function httpRequest(options, postData) {
  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => {
        data += chunk;
      });
      res.on('end', () => {
        try {
          resolve(JSON.parse(data));
        } catch (e) {
          resolve(data);
        }
      });
    });

    req.on('error', (err) => {
      reject(err);
    });

    if (postData) {
      req.write(postData);
    }

    req.end();
  });
}

// Crawl starten
async function startCrawl() {
  const options = {
    hostname: 'api.firecrawl.dev',
    port: 443,
    path: '/v1/crawl',
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${FIRECRAWL_API_KEY}`,
      'Content-Type': 'application/json'
    }
  };

  const postData = JSON.stringify({
    url: WEBSITE_URL,
    limit: 100,
    scrapeOptions: { formats: ['markdown'] }
  });

  try {
    const response = await httpRequest(options, postData);
    const crawlId = response.id;

    if (!crawlId) {
      console.error('Fehler: Crawl konnte nicht gestartet werden');
      console.error(JSON.stringify(response, null, 2));
      process.exit(1);
    }

    console.log(`Crawl ID: ${crawlId}`);
    return crawlId;
  } catch (err) {
    console.error('Fehler beim Starten des Crawls:', err.message);
    process.exit(1);
  }
}

// Status prüfen
async function checkStatus(crawlId) {
  const options = {
    hostname: 'api.firecrawl.dev',
    port: 443,
    path: `/v1/crawl/${crawlId}`,
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${FIRECRAWL_API_KEY}`
    }
  };

  while (true) {
    try {
      const response = await httpRequest(options);
      const status = response.status || 'unknown';
      console.log(`Status: ${status}`);

      if (status === 'completed') {
        const dateStr = new Date().toISOString().slice(0, 10).replace(/-/g, '');
        const outputFile = path.join(OUTPUT_DIR, `${dateStr}_crawl.json`);
        fs.writeFileSync(outputFile, JSON.stringify(response, null, 2));
        console.log(`Gespeichert in ${OUTPUT_DIR}`);
        break;
      } else if (status === 'failed') {
        console.error('Crawl fehlgeschlagen');
        process.exit(1);
      }

      // 5 Sekunden warten
      await new Promise(resolve => setTimeout(resolve, 5000));
    } catch (err) {
      console.error('Fehler beim Prüfen des Status:', err.message);
      process.exit(1);
    }
  }
}

// Hauptfunktion
(async () => {
  try {
    const crawlId = await startCrawl();
    await checkStatus(crawlId);
  } catch (err) {
    console.error('Unerwarteter Fehler:', err.message);
    process.exit(1);
  }
})();

File: python/websearch-crawl.py
#!/usr/bin/env python3
# websearch-crawl.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway1:scripts/websearch-crawl.sh
# auch in: OpenClaw@gateway2:scripts/websearch-crawl.sh
# Erzeugt: 2026-08-23 durch ABSTRACTIONS_MANAGER.py

# Web Search Script: Website Crawling mit Firecrawl
# Verwendung: ./websearch-crawl.py <URL> [OUTPUT_DIR]

import os
import sys
import json
import time
import subprocess
from datetime import datetime
import urllib.request
import urllib.error

def get_firecrawl_api_key():
    """API-Schlüssel aus Umgebungsvariable oder openclaw-Konfiguration"""
    key = os.environ.get('FIRECRAWL_API_KEY')
    if key:
        return key
    
    # Fallback auf openclaw.env Konfiguration
    try:
        config_path = os.path.expanduser('~/.openclaw/openclaw.env')
        with open(config_path, 'r') as f:
            for line in f:
                if line.startswith('OPENROUTER'):
                    # Extrahiere Wert zwischen Anführungszeichen
                    parts = line.split('"')
                    if len(parts) >= 2:
                        return parts[1]
    except FileNotFoundError:
        pass
    
    return None

def curl_post(url, headers, data):
    """Führt HTTP POST-Anfrage aus"""
    req = urllib.request.Request(url, data=data.encode('utf-8'), headers=headers)
    try:
        with urllib.request.urlopen(req) as response:
            return response.read().decode('utf-8')
    except urllib.error.HTTPError as e:
        return e.read().decode('utf-8')

def curl_get(url, headers):
    """Führt HTTP GET-Anfrage aus"""
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req) as response:
            return response.read().decode('utf-8')
    except urllib.error.HTTPError as e:
        return e.read().decode('utf-8')

def main():
    if len(sys.argv) < 2:
        print("Verwendung: {} <URL> [OUTPUT_DIR]".format(sys.argv[0]))
        sys.exit(1)
    
    website_url = sys.argv[1]
    output_dir = sys.argv[2] if len(sys.argv) > 2 else './crawled'
    firecrawl_api_key = get_firecrawl_api_key()
    
    if not firecrawl_api_key:
        print("FIRECRAWL_API_KEY muss gesetzt sein")
        sys.exit(1)
    
    os.makedirs(output_dir, exist_ok=True)
    print(f"Crawling {website_url}...")
    
    # Crawl starten
    headers = {
        "Authorization": f"Bearer {firecrawl_api_key}",
        "Content-Type": "application/json"
    }
    
    crawl_data = {
        "url": website_url,
        "limit": 100,
        "scrapeOptions": {"formats": ["markdown"]}
    }
    
    crawl_response_text = curl_post(
        "https://api.firecrawl.dev/v1/crawl",
        headers,
        json.dumps(crawl_data)
    )
    
    try:
        crawl_response = json.loads(crawl_response_text)
    except json.JSONDecodeError:
        print("Ungültige Antwort vom Server")
        print(crawl_response_text)
        sys.exit(1)
    
    crawl_id = crawl_response.get('id')
    if not crawl_id:
        print("Fehler: Crawl konnte nicht gestartet werden")
        print(json.dumps(crawl_response, indent=2))
        sys.exit(1)
    
    print(f"Crawl ID: {crawl_id}")
    
    # Status prüfen
    while True:
        status_response_text = curl_get(
            f"https://api.firecrawl.dev/v1/crawl/{crawl_id}",
            headers
        )
        
        try:
            status_response = json.loads(status_response_text)
        except json.JSONDecodeError:
            print("Ungültige Statusantwort vom Server")
            continue
        
        status = status_response.get('status', 'unknown')
        print(f"Status: {status}")
        
        if status == 'completed':
            timestamp = datetime.now().strftime('%Y%m%d')
            output_file = os.path.join(output_dir, f"{timestamp}_crawl.json")
            with open(output_file, 'w') as f:
                json.dump(status_response, f, indent=2)
            print(f"Gespeichert in {output_dir}")
            break
        elif status == 'failed':
            print("Crawl fehlgeschlagen")
            sys.exit(1)
        
        time.sleep(5)

if __name__ == '__main__':
    main()
